
Biostat 578 HW3
========================================================
Kenneth Chen

Assignment:

You will have to analyze the RNA-seq data presented in:
Henn, A. D. et al. High-resolution temporal response patterns to influenza vaccine reveal a distinct human plasma cell gene signature. Scientific Reports 3, 2327 (2013).
  
1. Get the data from GEO. Please look at the class lecture slides as we've done it already
2. Use voom and limma to find genes that are differentially expressed at each time point compared to baseline (day 0). Use an FDR cutoff of 0.01.
Display your results using pheatmap showing the log fold-change of the differentially expressed genes grouped by time point.  
3. Perform a GSEA analysis using camera and the MSigDB Reactome pathway gene signatures. Display your results using pheatmap, again group by timepoint. This is similar to what we've done in class.

Prior to running the current file ensure that these steps have been completed.
=======================================
1. Change working directory to source file directory by through Session -> Set Working Directory
2. Install necessary packages: GEOquery, limma, data.table, pheatmap, GSEABase
3. Create a sub-directory structure in the directory in which this file is stored of the form ./Data/GEO/GSE45735
4. Download relevant supplementary files by using this command
  getGEOSuppFiles("GSE45735", makeDirectory=FALSE, baseDir = "Data/GEO/")
5. Unzip and manually delete the last few lines of the T14 file. Offending rows will be clear upon visual inspection in a text editor that can recognize line breaks.
6. Recreate compressed T14 file in .gz format.
7. Download the necessary Reactome gene set file at http://www.broadinstitute.org/gsea/msigdb/download_file.jsp?filePath=/resources/msigdb/4.0/c2.cp.reactome.v4.0.symbols.gmt and save in subdirectory ./GSEA-sets


```{r message=FALSE}
#Load packages
library(GEOquery)
library(limma)
library(data.table)
library(pheatmap)
library(GSEABase)
```

Get the data from GEO. Please look at the class lecture slides as we've done it already
=======================================================================

```{r}
#Create a list of strings of the file names of the supplementary files. "full.names = TRUE" indicates that the path indicated in the call will be included in the strings.
files <- list.files(path = "Data/GEO/GSE45735", pattern = "GSE45735_T.*.gz", full.names = TRUE)

#Take the list of files denoted by files and convert each into a data.table. This command creates a list of data frames. Each data frame is created from one of the files whose name is in the "files" list. The header = TRUE designation turns the first row of those files into a column names.
file_list <- lapply(files, read.table, header=TRUE)

#There are sometimes more than one row per gene in these files. This is because reads can map to more than one gene. Remove duplicate rows.
#lapply(): apply the following function to each entry in "file_list"
#duplicated(): returns a boolean vector, default FALSE, TRUE for the second (and any more?) instance of a non-unique entry
#make the Gene column the row names, and remove that column
file_list_unique <- lapply(file_list, function(x){x<-x[!duplicated(x$Gene),]; 
                                                  x <- x[order(x$Gene),]; 
                                                  rownames(x) <- x$Gene;
                                                  x[,-1]})

#Create list of genes by finding the genes shared among all of the samples
gene_list <- Reduce(intersect, lapply(file_list_unique, rownames))
#subset data frames in file_list to only include genes in gene_list (shared genes)
file_list_unique <- lapply(file_list_unique, "[", gene_list,)
#merge all the data frames into a single data frame (column merge), and convert to matrix
matrix <- as.matrix(do.call(cbind, file_list_unique))

gd <- getGEO("GSE45735", destdir = "Data/GEO/")
pd <- pData(gd[[1]])
pd_small <- pd[!grepl("T13_Day8",pd$title),] #Day 8 data for T_13 patient not in file

#Create columns in pData for variables of interest: Day and subject (ie. patient)
pd_small$Day <- sapply(strsplit(gsub(" \\[PBMC\\]", "", pd_small$title),"_"),"[",2)
pd_small$subject <- sapply(strsplit(gsub(" \\[PBMC\\]", "", pd_small$title),"_"),"[",1)
colnames(matrix) <- rownames(pd_small) #GSM names to sample

new_set <- ExpressionSet(assayData = matrix+1) #add the matrix, now labeled with GSM numbers, to an expression set as the assay data 
pData(new_set) <- pd_small #assign the pdata created (cleaned from original data set) to that expression set

#This whole process is necessary because the GEO data for this experiment has no "assayData" values for the expression set created by the getGEO command, probably because it is an RNAseq data set (rather than a microarray as previous assignment). so: create new expression set from merge of data in supplementary files and pdata from getGEO
```

Use voom and limma to find genes that are differentially expressed at each time point compared to baseline (day 0). Use an FDR cutoff of 0.01. Display your results using pheatmap showing the log fold-change of the differentially expressed genes grouped by time point.  
==================================================================================

```{r fig.width = 14, fig.height = 12}
design <- model.matrix(~subject+Day, new_set)
new_set_voom <- voom(new_set,design = design)
lm <- lmFit(new_set_voom, design)
eb <- eBayes(lm)

FDR = 0.01 #set false discovery rate
days = 10
tt = vector("list", days)
for (i in 1:days) {
  coef = paste0("DayDay", i)
  tt_temp =  topTable(eb, coef=coef, number=Inf, sort.by="none")
  tt[[i]]=tt_temp
}

#there are no genes which meet FDR < 0.01 for all the days, so find set of genes which meets FDR < 0.01 on at least one day wrt baseline of Day 0
FDR = 0.01 #set false discovery rate
genes = dim(tt[[1]])[1]
keep = logical(genes)
for (i in 1:genes) {
  for (tt_temp in tt) {
    if (tt_temp$adj.P.Val[i] < 0.01) {
      keep[i] = TRUE
    }
  }
}

diffGenes = vector("list", days) #each element will be a vector of logFC values for genes where keep is TRUE (ie. there is one day were the FDR for that gene < 0.01), index corresponds to day, 
for (i in 1:days) {
  diffGenes[[i]] = tt[[i]]$logFC[keep]
}
diffGenes2 = do.call(cbind, diffGenes) #bind the list together into a single data frame, genes on rows, days on columns, elements are logFC
rownames(diffGenes2) = rownames(tt[[1]][keep,]) #get gene names
colnames(diffGenes2) = sapply(1:10, function(x) {paste0("Day ", x)}) # set column names to days

pheatmap(as.matrix(diffGenes2), cluster_cols = FALSE) #draw heatmap
```

Perform a GSEA analysis using camera and the MSigDB Reactome pathway gene signatures. Display your results using pheatmap, again group by timepoint. This is similar to what we've done in class.
=============================================================================

``` {r fig.width = 14, fig.height = 12} 
reactome_set <- getGmt("GSEA-sets/c2.cp.reactome.v4.0.symbols.gmt")
gene_ids <- geneIds(reactome_set)
# Camera requires gene-indices
sets_indices <- symbols2indices(gene_ids, rownames(new_set))

res <- vector("list",length = 10)
for(i in 1:10)
  {
  contrast <- paste0("DayDay",i)
  cont_matrix <- makeContrasts(contrast, levels=design)
  res[[i]] <- camera(new_set_voom, sets_indices, design=design, contrast=cont_matrix, sort=FALSE)
}

PValue <- sapply(res, function(x){ifelse(x$Direction=="Up", -10*log10(x$PValue), 10*log10(x$PValue))})
rownames(PValue) <- rownames(res[[1]])
PValue_max <- rowMax(abs(PValue))
PValue_small <- PValue[PValue_max>30, ]
anno <- data.frame(Time=paste0("Day",1:10))
rownames(anno) <- colnames(PValue_small)  <- paste0("Day",1:10)
pheatmap(PValue_small, cluster_cols=FALSE)
```